pymonetdb/filetransfer.py

```diff
@@ -0,0 +1,621 @@
+"""
+Classes related to file transfer requests as used by COPY INTO ON CLIENT.
+"""
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0.  If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.
+
+
+from abc import ABC, abstractmethod
+import codecs
+from importlib import import_module
+from io import BufferedIOBase, BufferedWriter, TextIOBase, TextIOWrapper
+from pathlib import Path
+from shutil import copyfileobj
+from typing import Optional, Union
+from pymonetdb import mapi
+from pymonetdb.exceptions import OperationalError, ProgrammingError
+
+
+def handle_file_transfer(mapi, cmd: str):
+    if cmd.startswith("r "):
+        parts = cmd[2:].split(' ', 2)
+        if len(parts) == 2:
+            try:
+                n = int(parts[0])
+            except ValueError:
+                pass
+            return handle_upload(mapi, parts[1], True, n)
+    elif cmd.startswith("rb "):
+        return handle_upload(mapi, cmd[2:], False, 0)
+    elif cmd.startswith("w "):
+        return handle_download(mapi, cmd[2:], True)
+    elif cmd.startswith("wb "):
+        return handle_download(mapi, cmd[3:], False)
+    else:
+        pass
+    # we only reach this if decoding the cmd went wrong:
+    mapi._putblock(f"Invalid file transfer command: {cmd!r}")
+
+
+def handle_upload(mapi, filename, text_mode, offset):
+    if not mapi.uploader:
+        mapi._putblock("No upload handler has been registered with pymonetdb\n")
+        return
+    skip_amount = offset - 1 if offset > 0 else 0
+    upload = Upload(mapi)
+    try:
+        try:
+            mapi.uploader.handle_upload(upload, filename, text_mode, skip_amount)
+        except Exception as e:
+            # We must make sure the server doesn't think this is a succesful upload.
+            # The protocol does not allow us to flag an error after the upload has started,
+            # so the only thing we can do is kill the connection
+            upload.error = True
+            mapi._sabotage()
+            raise e
+        if not upload.has_been_used():
+            raise ProgrammingError("Upload handler didn't do anything")
+    finally:
+        upload.close()
```

Comment:
a double try smells like something that can be solved more elegantly, but I don't understand good enough what is going on here yet.

Reply:
Probably a leftover from earlier versions.. I've removed the double try.